engine/src/main/java/org/hibernate/validator/internal/metadata/aggregated/CascadingMetaDataBuilder.java

```diff
@@ -211,12 +195,57 @@ public CascadingMetaDataBuilder merge(CascadingMetaDataBuilder otherCascadingTyp
 	public CascadingMetaData build(ValueExtractorManager valueExtractorManager, Object context) {
 		validateGroupConversions( context );
 
-		if ( containerElementTypesCascadingMetaData.isEmpty() ) {
+		// we are 100% sure that we have a container and its elements are marked for cascading
+		if ( !containerElementTypesCascadingMetaData.isEmpty() && hasContainerElementsMarkedForCascading ) {
+			return ContainerCascadingMetaData.of( valueExtractorManager, this, context );
```

Comment:
Wondering how we would react to:

```java
@Valid
private Map<@Valid KeyBean, ValueBean> map;
```

The spec advises you to not use it but because there is a risk for the values to be validated twice.

Reply:
Maybe we should have all this logic like this:

```java
if ( !cascading ) {
    if ( !containerElementTypesCascadingMetaData.isEmpty() && hasContainerElementsMarkedForCascading ) {
         // [...]
    }
    else {
       return NonContainerCascadingMetaData.of( this, context );
    }
}
```

Reply:
And we obviously need a new test case.